lua/luasnip/extras/snippet_list.lua

```diff
@@ -0,0 +1,92 @@
+local function get_context(snippet)
+	return {
+		name = snippet.name,
+		trigger = snippet.trigger,
+		description = snippet.dscr,
+		wordTrig = snippet.wordTrig and true or false,
+		regTrig = snippet.regTrig and true or false,
+		docstring = snippet:get_docstring(),
+	}
+end
+
+local function get_name(bufnr)
+	return "LuaSnip://Snippets buf " .. bufnr
+end
+
+local function printer(snippets)
+	return vim.inspect(snippets)
+end
+
+local win_opts = { foldmethod = "indent" }
+local buf_opts = { filetype = "lua" }
+
+local function set_win_opts(win, opts)
+	for opt, val in pairs(opts) do
+		vim.api.nvim_win_set_option(win, opt, val)
+	end
+end
+
+local function set_buf_opts(buf, opts)
+	for opt, val in pairs(opts) do
+		vim.api.nvim_buf_set_option(buf, opt, val)
+	end
+end
+
+local function make_scratch_buf(buf)
+	local opts = {
+		buftype = "nofile",
+		bufhidden = "wipe",
+		buflisted = false,
+		swapfile = false,
+		modified = false,
+		modeline = false,
+	}
+
+	for opt, val in pairs(opts) do
+		vim.api.nvim_buf_set_option(buf, opt, val)
+	end
+end
+
+local function open_snippet_list(opts)
+	opts = opts or {}
+	opts.get_name = opts.get_name or get_name
+	opts.printer = opts.printer or printer
+	opts.win_opts = opts.win_opts or win_opts
+	opts.buf_opts = opts.buf_opts or buf_opts
+
+	-- load snippets before changing windows/buffers
+	local available = require("luasnip").available
+	local snippets = available({ get_context = get_context })
+
+	-- get buffer handle before switching
+	local from_buf = vim.api.nvim_get_current_buf()
+
+	-- create and open buffer on right vertical split
+	vim.cmd("botright vnew")
```

Comment:
WDYT about making this configurable? I feel like a function would be appropriate, just a vimscript-string might be too restrictive.
(not sure about name though, `open`? `open_new`?)

Reply:
Hmm, so if I'm understanding you correctly you want the user to have control over the window that is opened. So there is ```vim.api.nvim_open_win``` which has a wholleee lot of config options for the new window. We could default to ```vim.cmd("bothright vnew")``` and if the user wants to customize anything they could just pass whatever they would pass to ```vim.api.nvim_open_win``` to make the window however they want.

What do you think?

Reply:
I think it might be more straightforward if we just say that "after `open`, the window (and buffer) the snippets should be displayed in, should both be current"

Another idea: we can pretty cleanly separate the steps we have to take here:
1. get info on available snippets (customize via `get_context`)
2. turn info into string (customize via `printer`)
3. display results of `printer` in some buffer (customize (currently) via `get_name`, `win_opts`, `buf_opts`).

We'd theoretically only need one option for the last one (to make it more abstract/capable), something like `display(printer_result, associated_bufnr)`, which could then open a split, a floating window, a new tab, add some weird highlighting to the buffer, whatever.

We could provide a default, which is basically just the code we have right now, and that itself (!!) could take the `get_name`, `win_opts`, `buf_opts`-parameters.
That should make this whole module very extensible, we could even implement a much more advanced `display` that collapses filetypes, has some interesting keybindings, ..., on this skeleton. (for such an advanced display we'd probably need some structured data from `printer`, so we should not restrain its result to `string`)

WDYT? Am I overcomplicating things? xD

Reply:
hahah I mean the original use case for myself was just to have a nicer way for me to search through the snippets to find the one I was looking for. Having all the customization is really awesome don't get me wrong but a bit overkill from my POV 😂. 

But from what I'm understanding we're not even changing the code that much just reorganizing a bit. You want to separate the ```open_snippet_list``` stuff into separate functions? Is that right?

I'm not sure what you want to do about the ```vim.cmd("botright vnew")```

> "after open, the window (and buffer) the snippets should be displayed in, should both be current"

That's what is happening already no? ```vim.cmd("botright vnew")``` opens a new vertical split with a buffer and makes both of those the current ones (Window and buffer).

You want to do something like this?

```lua
function open()
    vim.cmd("botright vnew")
    return vim.api.nvim_get_current_buf(), vim.api.nvim_get_current_win()
end
```

> Another idea: we can pretty cleanly separate the steps we have to take here:
    1. get info on available snippets (customize via get_context)
    2. turn info into string (customize via printer)
    3. display results of printer in some buffer (customize (currently) via get_name, win_opts, buf_opts).

```lua
function open_snippet_list(opts)
--- set opts (get_name, printer, win_opts, buf_opts)

-- Get info
local snippets = available(snip_info)

-- Stringify
local res  = printer(snippets)

-- Display
display(res, vim.api.nvim_get_current_buf())
-- In this function we would open the new window/buffer, set options, etc...
end
```

Something like that you mean? Or am I completely wrong?

Reply:
> Having all the customization is really awesome don't get me wrong but a bit overkill from my POV :joy:

xD Yeah, I also felt like it might be too much, but it's a really nice design, and we have lots of flexibility.
If you'd like to just get this done, I totally understand, we don't need all that now, can easily extend

> That's what is happening already no?

Yes! I wanted to propose we use that as the definition for what `open` should do, but having `open` return the buffer+window as you suggest seems better (more standard, if you know what I mean).

> But from what I'm understanding we're not even changing the code that much just reorganizing a bit. You want to separate the `open_snippet_list` stuff into separate functions?

Yes, this would make it easier to replace just one of the parts (as a user, and for us, since we just have to provide a different set of functions, and those can then be plugged into `open_snippet_list` as `opts`)
My idea would be
```lua
function default_open_from_opts(opts) -- with opts get_name, win_opts, buf_opts
   return function(printer_result)
       <basically what we currently use for opening the buffer, setting the buffer-options, ...>
   end
end

function open_snippet_list(opts) ... end -- with opts get_info(snippet), printer(snip_info_by_ft_snippet_type), open(printer_result)
```

I think that might be better because it doesn't restrict how the snippets are displayed in any way, while also allowing the more straightforward customization we have currently (don't have to re-implement the whole default_open to just tweak a window-option)

Reply:
Check out https://github.com/L3MON4D3/LuaSnip/pull/652/commits/1bc38c70210e6be08a51ea850aab926736e8f590!